Docker does not start when there is more than one default route

If you have an environment with multiple default routes, docker fails to start.

## Used versions

I tried several versions and used to ubuntu packages. The latest two of them are mentioned below.

```plain
# docker -v
Docker version 1.2.0, build fa7b24f
# docker -v
Docker version 1.1.2, build d84a070
# cat /etc/apt/sources.list.d/docker.list 
deb http://get.docker.io/ubuntu docker main
```

## Used distribution

Debian Wheezy with backports

```plain
# cat /etc/debian_version 
7.6
```

## routing table

This setup is not really common, but we are using this for several years. You can have even more fun with this setup, if you mix in iptables and packet marking in the mangle table.

```plain
# ip route
default 
        nexthop via 192.168.0.254  dev eth0 weight 1
        nexthop via 192.168.1.254  dev eth1 weight 1
192.168.0.0/24 dev eth0  scope link  src 192.168.0.1 
192.168.1.0/24 dev eth1  scope link  src 192.168.1.1 
# ip rule
0:      from all lookup local 
32760:  from 192.168.0.1 lookup eth0_table 
32761:  from 192.168.1.1 lookup eth1_table 
32766:  from all lookup main 
32767:  from all lookup default 
```

You can build this by entering the following commands. You may consider reading some resources for this, because it's a little bit complicated until you get this thing to work.

```plain
# echo 98 eth0_table >> /etc/iproute2/rt_tables
# echo 99 eth1_table >> /etc/iproute2/rt_tables
...
# ip rule add from 192.168.0.1 table eth0_table
# ip rule add from 192.168.1.1 table eth1_table
...
# ip route add default via 192.168.0.254 dev eth0 table eth0_table
# ip route add 192.168.0.0/24 dev eth0 src 192.168.0.1 table eth0_table
# ip route add default via 192.168.1.254 dev eth1 table eth1_table
# ip route add 192.168.1.0/24 dev eth1 src 192.168.1.1 table eth1_table
...
# ip route add default scope global \
nexthop via 192.168.0.254 dev eth0 weight 1 \
nexthop via 192.168.1.254 dev eth1 weight 1 \
...
```

## Error

Docker fails to measure the MTU to use in this setup.

```plain
goroutine 24 [running]:
runtime.panic(0xb48200, 0x1174573)
        /usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
github.com/docker/docker/daemon.GetDefaultNetworkMtu(0xae6260)
```

## Complete traceback

```plain
2014/08/29 12:30:56 docker daemon: 1.2.0 fa7b24f; execdriver: native; graphdriver: 
[27e626d9] +job serveapi(unix:///var/run/docker.sock)
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0x449eb2]

goroutine 24 [running]:
runtime.panic(0xb48200, 0x1174573)
        /usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
github.com/docker/docker/daemon.GetDefaultNetworkMtu(0xae6260)
        /go/src/github.com/docker/docker/daemon/config.go:67 +0x32
github.com/docker/docker/daemon.NewDaemonFromDirectory(0x1196b00, 0xc2080bde40, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/daemon/daemon.go:677 +0x76
github.com/docker/docker/daemon.NewDaemon(0x1196b00, 0xc2080bde40, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/daemon/daemon.go:666 +0x43
main.func·001()
        /go/src/github.com/docker/docker/docker/daemon.go:44 +0x43
created by main.mainDaemon
        /go/src/github.com/docker/docker/docker/daemon.go:56 +0x17c

goroutine 16 [chan receive]:
github.com/docker/docker/api/server.ServeApi(0xc208052700, 0x7f30786e0430)
        /go/src/github.com/docker/docker/api/server/server.go:1361 +0x473
github.com/docker/docker/engine.(*Job).Run(0xc208052700, 0x0, 0x0)
        /go/src/github.com/docker/docker/engine/job.go:79 +0x7d0
main.mainDaemon()
        /go/src/github.com/docker/docker/docker/daemon.go:78 +0x56c
main.main()
        /go/src/github.com/docker/docker/docker/docker.go:54 +0x2b1

goroutine 19 [finalizer wait]:
runtime.park(0x41bf90, 0x1193b80, 0x1179889)
        /usr/local/go/src/pkg/runtime/proc.c:1369 +0x89
runtime.parkunlock(0x1193b80, 0x1179889)
        /usr/local/go/src/pkg/runtime/proc.c:1385 +0x3b
runfinq()
        /usr/local/go/src/pkg/runtime/mgc0.c:2644 +0xcf
runtime.goexit()
        /usr/local/go/src/pkg/runtime/proc.c:1445

goroutine 20 [syscall]:
os/signal.loop()
        /usr/local/go/src/pkg/os/signal/signal_unix.go:21 +0x1e
created by os/signal.init·1
        /usr/local/go/src/pkg/os/signal/signal_unix.go:27 +0x32

goroutine 17 [syscall]:
runtime.goexit()
        /usr/local/go/src/pkg/runtime/proc.c:1445

goroutine 23 [chan receive]:
github.com/docker/docker/pkg/signal.func·002()
        /go/src/github.com/docker/docker/pkg/signal/trap.go:29 +0x88
created by github.com/docker/docker/pkg/signal.Trap
        /go/src/github.com/docker/docker/pkg/signal/trap.go:53 +0x28f

goroutine 25 [runnable]:
github.com/docker/docker/api/server.func·008()
        /go/src/github.com/docker/docker/api/server/server.go:1354
created by github.com/docker/docker/api/server.ServeApi
        /go/src/github.com/docker/docker/api/server/server.go:1357 +0x3fa
```

## Conclusion

I've tried to supply a patch for this issue, but this is beyond something we can call a quick fix. It might break the internal data structure at this point. This is because all layers of code between this exception and the real method which does measure the MTU are designed to have only one value eg. one default route in place.

Perhaps you can use the least MTU or something like this.

On the other hand, as mentioned above, there are several ways to manipulate packets going through the routing tables. So you may consider that this value can be set and is not discovered. Because you cannot determinate the outgoing route for every packet when docker starts. Even if you run this discovery every second, you'll miss some routes, as this decision can be made for every single packet with various results.

--- a/daemon/container_operations_windows.go
+++ b/daemon/container_operations_windows.go
@@ -162,10 +162,6 @@ func detachMounted(path string) error {
 	return nil
 }
 
-func getDefaultRouteMtu() (int, error) {
-	return -1, errSystemNotSupported
-}
-
 func killProcessDirectly(container *container.Container) error {
 	return nil
 }

--- a/daemon/daemon.go
+++ b/daemon/daemon.go
@@ -1382,12 +1382,7 @@ func setDefaultMtu(config *Config) {
 		return
 	}
 	config.Mtu = defaultNetworkMtu
-	if routeMtu, err := getDefaultRouteMtu(); err == nil {
-		config.Mtu = routeMtu
-	}
 }
-
-var errNoDefaultRoute = errors.New("no default route was found")
 
 // verifyContainerSettings performs validation of the hostconfig and config
 // structures.

--- a/daemon/daemon_unix.go
+++ b/daemon/daemon_unix.go
@@ -33,7 +33,6 @@ import (
 	"github.com/docker/libnetwork/types"
 	blkiodev "github.com/opencontainers/runc/libcontainer/configs"
 	"github.com/opencontainers/runc/libcontainer/label"
-	"github.com/vishvananda/netlink"
 )
 
 const (
@@ -668,25 +667,6 @@ func (daemon *Daemon) conditionalMountOnStart(container *container.Container) er
 // during the cleanup of a container to unmount.
 func (daemon *Daemon) conditionalUnmountOnCleanup(container *container.Container) {
 	daemon.Unmount(container)
-}
-
-// getDefaultRouteMtu returns the MTU for the default route's interface.
-func getDefaultRouteMtu() (int, error) {
-	routes, err := netlink.RouteList(nil, 0)
-	if err != nil {
-		return 0, err
-	}
-	for _, r := range routes {
-		// a nil Dst means that this is the default route.
-		if r.Dst == nil {
-			i, err := net.InterfaceByIndex(r.LinkIndex)
-			if err != nil {
-				continue
-			}
-			return i.MTU, nil
-		}
-	}
-	return 0, errNoDefaultRoute
 }
 
 func restoreCustomImage(driver graphdriver.Driver, is image.Store, ls layer.Store, ts tag.Store) error {

--- a/docs/userguide/networking/default_network/custom-docker0.md
+++ b/docs/userguide/networking/default_network/custom-docker0.md
@@ -16,7 +16,7 @@ The information in this section explains how to customize the Docker default bri
 
 By default, the Docker server creates and configures the host system's `docker0` interface as an _Ethernet bridge_ inside the Linux kernel that can pass packets back and forth between other physical or virtual network interfaces so that they behave as a single Ethernet network.
 
-Docker configures `docker0` with an IP address, netmask and IP allocation range. The host machine can both receive and send packets to containers connected to the bridge, and gives it an MTU -- the _maximum transmission unit_ or largest packet length that the interface will allow -- of either 1,500 bytes or else a more specific value copied from the Docker host's interface that supports its default route.  These options are configurable at server startup:
+Docker configures `docker0` with an IP address, netmask and IP allocation range. The host machine can both receive and send packets to containers connected to the bridge, and gives it an MTU -- the _maximum transmission unit_ or largest packet length that the interface will allow -- of 1,500 bytes. These options are configurable at server startup:
 - `--bip=CIDR` -- supply a specific IP address and netmask for the `docker0` bridge, using standard CIDR notation like `192.168.1.5/24`.
 
 - `--fixed-cidr=CIDR` -- restrict the IP range from the `docker0` subnet, using the standard CIDR notation like `172.167.1.0/28`. This range must be an IPv4 range for fixed IPs (ex: 10.20.0.0/16) and must be a subset of the bridge IP range (`docker0` or set using `--bridge`). For example with `--fixed-cidr=192.168.1.0/25`, IPs for your containers will be chosen from the first half of `192.168.1.0/24` subnet.

